Sources/Datasets/datasets/image/MNIST.swift

```diff
@@ -1,99 +1,108 @@
+        get {
+            return MNISTDataLoader(split: .train)
+        }
+    }
+    public static var test: MNISTDataLoader {
+        get {
+            return MNISTDataLoader(split: .test)
+        }
+    }
+
+    public static var info = mnistInfo
+
+    private init() {}
 }
 
-fileprivate let MNISTInfo = S5TFDatasetInfo(
+fileprivate let mnistInfo = S5TFDatasetInfo(
     name: "mnist",
     version: "0.0.1",
     description: "The MNIST database of handwritten digits. 60000 train examples and 10000 test examples with image and label features.",
-    homepage: "http://yann.lecun.com/exdb/mnist/",
-    citation: """
-    @article{lecun2010mnist,
-      title={MNIST handwritten digit database},
-      author={LeCun, Yann and Cortes, Corinna and Burges, CJ},
-      journal={ATT Labs [Online]. Available: http://yann.lecun.com/exdb/mnist},
-      volume={2},
-      year={2010}
-    }
-    """)
-
-public struct MNISTIterator: S5TFDatasetIterator {
-    typealias _Dataset = MNIST // swiftlint:disable:this type_name
-    private var dataset: MNIST
+    homepage: URL(string: "http://yann.lecun.com/exdb/mnist/")!
+)
 
+public struct MNISTDataLoader: S5TFDataLoader {
     private var index = 0
 
-    public init(dataset: MNIST) {
-        self.dataset = dataset
-    }
-
-    mutating public func next() -> MNISTBatch? {
-        guard index < dataset.data.count else { return nil }
-        let data = dataset.data[index]
-        index += 1
-        return MNISTBatch(data: data[0], labels: data[1])
-    }
-}
+    let split: S5TFSplit
+    let batchSize: Int
+    let data: Tensor<UInt8>
+    let labels: Tensor<Int64>
 
-public struct MNIST: S5TFDataset {
-    typealias _Iterator = MNISTIterator // swiftlint:disable:this type_name
-
-    internal var data = [[1, 2], [3, 4], [5, 6]]
-
-    public func batched(_ batchSize: Int) -> MNISTIterator {
-        return MNISTIterator(dataset: self)
-    }
-
-    public var train: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .train)
-    }
-    public var validation: MNIST {
-        fatalError("The split does not exist for this dataset.")
-    }
-    public var test: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .test)
-    }
-    public var all: MNIST {
-        guard self.split == .undefined else { fatalError("This property can only be accessed from an undefined split.") }
-        return MNIST(split: .all)
-    }
-
-    public let split: S5TFSplit
-
-    public init() { self.split = .undefined }
-    private init(split: S5TFSplit) {
+    internal init(split: S5TFSplit, batchSize: Int) {
+        self.batchSize = batchSize
         self.split = split
-        let downloader = Downloader()
+
+        let semaphore = DispatchSemaphore(value: 0)
```

Comment:
You are only using one semaphore, does that work? You could increment the value if it doesn't.

Reply:
Sorry, not very familiar with the semaphore concept, what does the value entail?

Reply:
Check [this](https://developer.apple.com/documentation/dispatch/dispatchsemaphore). I'm more than happy to answer any questions you might have.